include/Transforms/Functional.h

```diff
@@ -0,0 +1,487 @@
+
+  detail::SimpleRewriter rewriter(scope->getContext());
+  // A list of all the results.
+  SmallVector<typename Unpack::type> results;
+  scope->walk([pattern = std::forward<PatternT>(pattern), &rewriter, &results,
+               &args...](Operation *op) {
+    rewriter.setInsertionPoint(op);
+    auto result = pattern(op, rewriter, std::forward<Args>(args)...);
+    // If the pattern applied, unpack the result and store it.
+    if (succeeded(result))
+      results.push_back(Unpack::unpack(std::move(result)));
+  });
+  return results;
+}
+
+/// Apply a pattern directly on an operation, for each operation in a list.
+template <typename ListT, typename PatternT, typename... Args,
+          bool = PatternConcept<PatternT>::verify()>
+auto applyForEach(ListT &&list, PatternT &&pattern, Args &&...args) {
+  using Traits = llvm::function_traits<std::decay_t<PatternT>>;
+  using Unpack = detail::UnpackFailureOr<typename Traits::result_t>;
+
+  // A list of all the results.
+  SmallVector<typename Unpack::type> results;
+  for (auto op : list) {
+    auto result = applyAt(op, std::forward<PatternT>(pattern),
+                          std::forward<Args>(args)...);
+    // The pattern applied, unpack the result and store it.
+    if (succeeded(result))
+      results.push_back(Unpack::unpack(std::move(result)));
+  }
+  return results;
+}
+
+/// Find the nearest proper ancestor that is isolated from above.
+Operation *findNearestScope(Operation *op);
+
+/// Clone the regions of an operation.
+SmallVector<std::unique_ptr<Region>> cloneRegionsOf(Operation *op);
+/// Replace the regions of an operation with the given regions.
+void takeRegions(Operation *op, ArrayRef<std::unique_ptr<Region>> regions);
+
+namespace detail {
+template <typename OpT, typename SequenceT, typename... Args,
+          bool = PatternConcept<SequenceT>::verify()>
+auto applySequence(PatternRewriter &rewriter, Operation *scope, OpT op,
+                   SequenceT &&sequence, Args &&...args) {
+  // Save a copy of the scope.
+  SmallVector<std::unique_ptr<Region>> backup = cloneRegionsOf(scope);
+
+  rewriter.setInsertionPoint(op);
+  auto result = std::forward<SequenceT>(sequence)(op, rewriter,
+                                                  std::forward<Args>(args)...);
+  // If the match failed, rewind the scope.
+  if (failed(result))
+    takeRegions(scope, backup);
+  return result;
+}
+} // end namespace detail
+
+/// Apply a pattern sequence directly on an operation. If any pattern in the
+/// sequence fails, then the state of the IR is restored. Patterns in the
+/// sequence should not query or modify IR outside the nearest scope.
+/// TODO: The rollback is not efficient. All of the IR underneath the nearest
+/// scope is copied.
+template <typename OpT, typename SequenceT, typename... Args,
+          bool = PatternConcept<SequenceT>::verify()>
+auto applySequenceAt(OpT op, SequenceT &&sequence, Args &&...args) {
+  Operation *scope = findNearestScope(op);
```

Comment:
Same comment re ScopeOp vs Operation* and future with interfaces.
But I won't insist anymore if you fell more strongly here :)

Reply:
I'll introduce an interface